lib/block_view_association_test.go

```diff
@@ -0,0 +1,1866 @@
+		require.Zero(t, count)
+
+		// Failed query: AssociationValue and AssociationValuePrefix specified
+		userAssociationQuery = &UserAssociationQuery{
+			TargetUserPKID:         m3PKID,
+			AssociationType:        "ENDORSEMENT",
+			AssociationValue:       "C#",
+			AssociationValuePrefix: "C",
+		}
+		userAssociationEntries, err = utxoView().GetUserAssociationsByAttributes(userAssociationQuery)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "invalid query params")
+		require.Nil(t, userAssociationEntries)
+
+		count, err = utxoView().CountUserAssociationsByAttributes(userAssociationQuery)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "invalid query params")
+		require.Zero(t, count)
+
+		// Failed query if Badger: no Transactor or TargetUser specified
+		userAssociationQuery = &UserAssociationQuery{
+			AssociationType:  "ENDORSEMENT",
+			AssociationValue: "C#",
+		}
+		userAssociationEntries, err = utxoView().GetUserAssociationsByAttributes(userAssociationQuery)
+		if chain.postgres != nil {
+			require.NoError(t, err)
+			require.Len(t, userAssociationEntries, 1)
+			require.Equal(t, userAssociationEntries[0].TargetUserPKID, m3PKID)
+		} else {
+			require.Error(t, err)
+			require.Contains(t, err.Error(), "invalid query params")
+			require.Nil(t, userAssociationEntries)
+		}
+		count, err = utxoView().CountUserAssociationsByAttributes(userAssociationQuery)
+		if chain.postgres != nil {
+			require.NoError(t, err)
+			require.Equal(t, count, uint64(1))
+		} else {
+			require.Error(t, err)
+			require.Contains(t, err.Error(), "invalid query params")
+			require.Zero(t, count)
+		}
+
+		// Failed query if Badger: empty AssociationType and non-empty AssociationValue
+		userAssociationQuery = &UserAssociationQuery{
+			TransactorPKID:   m4PKID,
+			AssociationValue: "C#",
+		}
+		userAssociationEntries, err = utxoView().GetUserAssociationsByAttributes(userAssociationQuery)
+		if chain.postgres != nil {
+			require.NoError(t, err)
+			require.Len(t, userAssociationEntries, 1)
+			require.Equal(t, userAssociationEntries[0].TargetUserPKID, m3PKID)
+		} else {
+			require.Error(t, err)
+			require.Contains(t, err.Error(), "invalid query params")
+			require.Nil(t, userAssociationEntries)
+		}
+		count, err = utxoView().CountUserAssociationsByAttributes(userAssociationQuery)
+		if chain.postgres != nil {
+			require.NoError(t, err)
+			require.Equal(t, count, uint64(1))
+		} else {
+			require.Error(t, err)
+			require.Contains(t, err.Error(), "invalid query params")
+			require.Zero(t, count)
+		}
+
+		// Failed query if Badger: non-empty AssociationTypePrefix and non-empty AssociationValue
+		userAssociationQuery = &UserAssociationQuery{
+			TransactorPKID:        m4PKID,
+			AssociationTypePrefix: "ENDORSE",
+			AssociationValue:      "C#",
+		}
+		userAssociationEntries, err = utxoView().GetUserAssociationsByAttributes(userAssociationQuery)
+		if chain.postgres != nil {
```

Comment:
My current plan is actually for these to diverge. This is only for query logic not used in consensus. But Postgres has fewer limitations than Badger when it comes to querying due to the limited set of Badger key indices we create. Postgres allows for virtually any combination of columns to query. For that reason, a Postgres backend would actually allow certain queries that a Badger backend would fail on. IMO that's ok as this is not used in consensus and to artificially restrict Postgres query functionality just for parity seems unneccessarily handicapping. But LMK what you think.

Reply:
I see - yeah that's kinda weird. I think it's okay for now and eventually we'll move query support out of core which will get rid of this divergence.